Modules/IP/Sources/IP/IP.swift

```diff
@@ -18,56 +27,347 @@ public enum IPMode {
+        self.init(address: address)
+    }
+}
+
+extension IP {
+    private static func assertValidPort(_ port: Int) throws {
         if port < 0 || port > 0xffff {
             throw IPError.invalidPort
         }
     }
+
+    fileprivate static func ip(address: String, port: Int, mode: IPMode, deadline: Double) throws -> Address {
+        do {
+            return try IP.ip(address: address, port: port, mode: mode)
+        } catch {
+            return try IP.remoteIP(address: address, port: port, mode: mode, deadline: deadline)
+        }
+    }
+
+    static private var dns_conf: UnsafeMutablePointer<dns_resolv_conf> = {
+        var rc: Int32 = 0
+        return dns_resconf_local(&rc)!
+    }()
+
+    static private var dns_hosts: OpaquePointer = {
+        var rc: Int32 = 0
+        return dns_hosts_local(&rc)!
+    }()
+
+    static private var dns_hints: OpaquePointer = {
+        var rc: Int32 = 0
+        return dns_hints_local(dns_conf, &rc)!
+    }()
+
+    private static func remoteIP(address: String, port: Int, mode: IPMode, deadline: Double) throws -> Address {
+        try IP.assertValidPort(port)
+        var rc: Int32 = 0
+        var dns_opts = dns_options()
+        let resolver = dns_res_open(dns_conf, dns_hosts, dns_hints, nil, &dns_opts, &rc)
+        var hints = addrinfo()
+        hints.ai_family = PF_UNSPEC
+        let ai = dns_ai_open(address, String(port), DNS_T_A, &hints, resolver, &rc)!
+        defer { dns_ai_close(ai) }
+        dns_res_close(resolver)
+
+        var ipv4: UnsafeMutablePointer<addrinfo>? = nil
+        var ipv6: UnsafeMutablePointer<addrinfo>? = nil
+        var it: UnsafeMutablePointer<addrinfo>? = nil
+
+        loop: while true {
+            rc = withUnsafeMutablePointer(to: &it) {
+                dns_ai_nextent($0, ai)
+            }
+
+            switch rc {
+            case EAGAIN:
+                let fd = dns_ai_pollfd(ai)
+                do {
+                    try poll(fd, events: .read, deadline: deadline)
+                    // TODO: export this in Venice
+                    fdclean(fd)
+                    continue loop
+                } catch PollError.timeout {
+                    throw DNSError.timeout
+                }
+            case ENOENT:
+                break loop
+            default:
+                if ipv4 == nil,
+                    let it = it,
+                    it.pointee.ai_family == AF_INET {
+                        ipv4 = it
+                } else if ipv6 == nil,
+                    let it = it,
+                    it.pointee.ai_family == AF_INET6 {
+                        ipv6 = it
+                } else {
+                    free(it)
+                }
+
+                if ipv4 != nil && ipv6 != nil {
+                    break loop
+                }
+            }
+        }
+
+        switch mode {
+        case .ipV4:
+            if ipv6 != nil {
+                free(ipv6)
```

Comment:
capacity is 1 in this case (you would have more than one if you have a list).

Deinitialize just sets the memory to zeros iirc, so no you don't need to.